### PR TITLE
Add pre-commit config with a few hooks and mdformat

### PR DESCRIPTION
**Pull request by lysnikolaou**

### CODE REVIEW

**Comment by lysnikolaou** [comment]
This is ready to be reviewed. What do you think @rgommers @ngoldbaum @Fidget-Spinner @andfoy? Are you happy with these hooks/checks?

**Comment by rgommers** [comment]
I'm honestly not sure. The main changes to C code are:

- `PYMODINIT_FUNC` which is made less idiomatic (all examples in the CPython code have it as two lines as well, e.g.,: https://docs.python.org/3/extending/extending.html)
- adding a lot of spacing under `PyModuleDef_Slot module_slots`, which looks worse
- making the `int do_modification` a one-liner, which looks a little worse
- having to declare it's not C but all C++, I guess to make the checker run.

The yaml changes look fine. Line breaks are sometimes worse, sometimes better. It's only a few changes, so I don't have strong feelings either way. I'll leave it to others, who will probably be writing more of the C code examples than me, to decide whether to merge.

File: docs/ci.md [hunk 1]
```diff
@@ -4,17 +4,17 @@ Currently the `setup-python` GitHub Action [does not
...
   free-threaded:
     runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@...
-    - uses: deadsnakes/action@...
-      with:
-          python-version: '3.13.0b2'
+      - uses: actions/checkout@...
+      - uses: deadsnakes/action@...
+        with:
```
Comment (ngoldbaum) [inline docs/ci.md:16]: For what it's worth, the NumPy github actions config doesn't have these tabs. I have no idea if adding them is a no-op. Hopefully it is but worth double-checking.

File: docs/ci.md [hunk 2]
```diff
@@ -31,15 +31,15 @@ You will need to specify the following variables in the environment for the
...
-          CIBW_BUILD_FRONTEND: "pip; args: --no-build-isolation"
+  - name: Build wheels
+    uses: pypa/cibuildwheel@...
+    env:
+      CIBW_PRERELEASE_PYTHONS: true
+      CIBW_FREE_THREADED_SUPPORT: true
+      CIBW_BUILD: cp313t-${{ matrix.buildplat }}
+      # TODO: remove along with installing build deps in
+      # cibw_before_build.sh when a released cython can build numpy
+      CIBW_BUILD_FRONTEND: 'pip; args: --no-build-isolation'
```
Comment (ngoldbaum) [inline docs/ci.md:42]: Similarly, it's worth double checking all these changes actually work. They probably do but I don't want to publish incorrect CI config snippets.

File: docs/porting.md [hunk 3]
```diff
@@ -92,14 +90,12 @@ C or C++ extension modules using multi-phase initialization can specify the
 [`Py_mod_gil`](https://docs.python.org/3.13/c-api/module.html#c.Py_mod_gil)
 module slot like this:
 
-```c
-static PyModuleDef_Slot module_slots[] = {
-    ...
+```cpp
```
Comment (ngoldbaum) [inline docs/porting.md:93]: I agree with Ralf it's unintuitive that we have to have C++ code snippets now.

File: docs/tracking.md [hunk 4]
```diff
@@ -18,31 +17,26 @@ hence this page does not aim to track pure Python packages.
...
+| scikit-build-core | :material-check-bold: | :material-check-bold: |           0.9.5            |                                                            |                                                                                          |
+| scikit-learn      | :material-check-bold: |                       |           1.6.0            |                       :simple-linux:                       | [:simple-anaconda:](https://anaconda.org/scientific-python-nightly-wheels/scikit-learn/) |
+| SciPy             | :material-check-bold: |                       |           1.15.0           |               :simple-linux: :simple-apple:                |    [:simple-anaconda:](https://anaconda.org/scientific-python-nightly-wheels/scipy/)     |
+| setuptools        | :material-check-bold: | :material-check-bold: |           69.5.0           |                                                            |                                                                                          |
+
+[^cmake]: Windows isn't correctly handled yet in CMake 3.30.0, see [cmake#26016](https://gitlab.kitware.com/cmake/cmake/-/issues/26016)
```
Comment (ngoldbaum) [inline docs/tracking.md:42]: why did the table get reformatted?

**Review by ngoldbaum — COMMENTED** [review]
If the bot is auto-fixing this stuff I don't mind much, but I don't think the changes are actually improving much, so maybe not worth the hassle?

File: docs/ci.md (hunk 1 above)
Comment (lysnikolaou) [reply docs/ci.md:16]: This is a no-op, yes, and more idiomatic as far as I understand. https://stackoverflow.com/a/71968274 for more context.

File: docs/ci.md (hunk 2 above)
Comment (lysnikolaou) [reply docs/ci.md:42]: https://cibuildwheel.pypa.io/en/stable/options/#free-threaded-support explicitly mentions that setting this to `true` works and goes on to give examples where the environment variable is set to 1.

About the string, the yaml file in the [`pandas` PR](https://github.com/pandas-dev/pandas/pull/59136/files) has quotes around it and it's okay.

File: docs/tracking.md (hunk 4 above)
Comment (lysnikolaou) [reply docs/tracking.md:42]: It's now all formatted and aligned in single lines, which I guess makes it a bit easier for a human to parse. At least it feels that way to me.

Before:
Screenshot 2024-07-19 at 15 30 23

After:
Screenshot 2024-07-19 at 15 30 54

Comment (ngoldbaum) [reply docs/tracking.md:42]: Nice I agree this is better, especially if your editor is configured not to line wrap.

**Comment by lysnikolaou** [comment]
Agreed with both your sentiments, I also have mixed feelings about this.

My proposal would be to get rid of `clang-format`, it's not a big improvement and if we want it to format the "right" way (like CPython/NumPy etc. do), we'll have to maintain a more complicated clang-format file which isn't really worth it.

I like having the rest in place, if only for not really caring about trailing whitespace, keeping formatting consistent in `.md` files etc.

File: docs/tracking.md (hunk 4 above)
Comment (lysnikolaou) [reply docs/tracking.md:42]: FWIW, if you're using VS Code on macOS, Option+Z toggles between line wrap mode.

**Comment by lysnikolaou** [comment]
I think this is ready for a final decision.

**Comment by ngoldbaum** [comment]
Another argument for doing this is that we can add codespell, which currently has a high signal-to-noise ratio on the docs:

```
$ codespell docs
docs/porting.md:179: occure ==> occur, occurred
```

(I'll fix the spelling issue in a separate PR)

**Comment by lysnikolaou** [comment]
Good point. Though `codespell` might not be able to always autofix (for example doing `codespell -w docs` does not auto-fix the "occure" spelling mistake).

**Review by rgommers — APPROVED** [review]
Looks like we have two in favor and no one against, so let's just give this a go! Thanks @lysnikolaou and @ngoldbaum.

**Pull request merged**